Step1

yeahg-dev opened the pull request:
안녕하세요 줄라이!
소중한 시간 내어 리뷰를 맡아주셔서 감사드립니다!😃
처음으로 MVVM 아키텍처와 rxSwift를 사용해서 많이 헤메었던것 같습니다.
진행하며 어려웠던 점과 고민되었 던 점을 정리해보았습니다. 
한 번 읽어보시고 조언 주시면 정말 감사하겠습니다!

아직 많이 부족하지만, 잘 부탁드립니다😊

# 구현 내용
- JuiceMaker뷰(쥬스 주문 화면), FruitStock뷰(과일 재고 수정 화면) UI 구현
-  앱 전체에서 과일재고를 공급하는 FruitStockRepository 싱글톤 객체 구현
- JuiceMaker뷰: 현재 재고를 Repository로부터 받아 뷰에 바인드, 쥬스 주문 버튼 탭시 과일 재고 수정 및 뷰 바인드 구현
- FruitStock뷰: 현재 재고를 Repository로부터 받아 뷰에 바인드, 스테퍼 액션으로 과일 재고 수정하는 로직 및 뷰 바인드 구현

### 미구현 기능 및 고쳐야할 점
- JuiceMaker뷰: 재료가 모자람에도 수파쥬스, 딸복쥬스 주문이 되는 버그 수정
- JuiceMaker뷰: 쥬스 주문 성공 및 실패 시 알럿 프레젠트 기능 구현
- FruitStock뷰: 재고 수정 오류에 대한 알럿 프레젠트 기능 구현

<br>

# 고민한 점
## Observable 사용
뷰 모델에서는 onNext로 아이템을 받아 뷰로 다시 아이템을 방출하거나, 뷰로 부터 아이템을 받아야하기 때문에 메서드에서 Observable을 리턴하는 것이 자연스럽게 느껴졌습니다. 
그런데 비지니스로직을 담당하는 모델이나 데이터를 처리하는 레포지토리에서는 언제 Observable을 리턴하고, 리턴하지 않아도 될지에 대한 고민이 되었습니다. 시행 착오 끝에 "메서드에 대한 결과가 스트림으로 계속 이어져야하는 경우에 Observable을 사용하면 용이하겠다" 라는 작은 결론을 지어보게되었습니다.😅
일전에 MVVM에 대해 설명해주실 때 Observable은 Repository에서 생성하고, 뷰에서 subscribe 해야한다. stream이 끊기면 안된다. 라고 하셨던 팁들이 생각이 났었는데요. 팁과 같이 구현해보려고 노력하였으나, 제가 아직 사용에 대한 이해가 부족하여 뷰모델과 비지니스로직에서도 subscribe를 해서 분기를 하거나 새로운 observable을 리턴하는 것과 같은 것들이 생겨나게되었습니다.ㅠㅠ

위와 같은 분기가 일어나게 된 배경은 아래와 같습니다.

뷰에서 이벤트를 받아 뷰모델이 비지니스 로직을 실행시 (예를 들어 쥬스 주문), 과일 재료가 충분하지 확인을 해야하고 이는 repository와의 소통이 필요합니다. 따라서 비지니스로직에서 repository에서 생성한 과일 재고observable을 구독하고, 쥬스 생성이 가능한지 판단 후 가능하다면 repository에서 과일 재고를 수정하고, 수정이 성공적으로 완료 되었다면 repository에서 과일재고를 읽어와서 observable로 뷰로 전달하도록 구현했습니다.  현재는 repository가 딕셔너리이지만, repository가 서버로 부터 데이터를 받아올 경우, 통신의 성공과 실패를 확인한 다음 결과에 따라 뷰에 보여줘야한다고 생각하여 오버스럽지만 이렇게 구현하게 되었습니다.

통신을 가정할 때, 하나의 스트림으로 자연스럽게 이어지는 로직을 구현하려면 Observable을 어떻게 활용하면 좋을지와 
Observable은 언제 리턴하시는지 줄라이의 경험과 의견이 궁금합니다😊

## 오류 처리에 대한 고민
위에서 말씀드린 Observable사용에 대한 고민과 이어지는 부분인데요. 재료가 있어서 쥬스제조에 성공한 케이스와 실패한 케이스 두가지 있을 때,
하나의 Observable로 뷰에 바인드 시키려하다보니, 실패한 경우 에러를 흘려보내게 되면 해당 스트림이 완전히 종료가 되어 다시 이벤트를 받지 못하는 경우가 있엇습니다. 

이에 대한 대안으로 비지니스 로직에서 성공과 실패에 대한 케이스를 타입으로 정의해서 Observable에 담아 흘려보내는 방법을 생각하게 되었습니다. 
```swift
  enum FruitStockModification {
        case success
        case deficientFruitStockFailure
}
```
그런데 이렇게 구현한 결과 뷰모델에서 분기처리를 해주어야 해서 인덴트가 많이 생기고, 가독성도 안좋아진 결과를 초래한 것 같습니다.
```swift
  input.strawberryStepperValueObservable?
            .map({[weak self] stepperValue in
                // FIXME: - initialStrawberryStock은 뷰컨트롤러에서 loadStock된 후 값이 생김
                (self?.initialStrawberryStock ?? 10) + Int(stepperValue)
            })
            .flatMap({ newValue in
                self.juiceMaker.modifiedFruitStockObservable(of: .strawberry, with: newValue)
            })
            .subscribe(
                onNext: { fruitStockModificationResult in
                    switch fruitStockModificationResult {
                    case .success:
                        // 재고 수정이 성공적으로 된 경우에만 레포지토리로부터 최신 재고를 읽어옴.
                        self.juiceMaker.fruitStockObservable(of: .strawberry)
                            .map{String($0)}
                            .subscribe(onNext: { stock in
                                self.strawberryStockObservable.onNext(stock)
                            })
                            .disposed(by: self.disposeBag)
                    case .deficientFruitStockFailure  :
                        self.notificationObservable.onNext(UserNotification())
                    }
                })
            .disposed(by: disposeBag)
```

이와 같이 비지니스로직 실행시 여러가지방향의 output을 보내줘야하는 경우,
에러를 내보내어 스트림을 종료시키지 않고 계속 진행시킬 수 있는 방법에 대해 조언주시면 감사하겠습니다! 

## 뷰모델이 커진 상황
과일의 스테퍼마다 흐름을 관찰하다보니 뷰모델의 코드가 엄청 많아졌습니다.
UI의 input이 많은 경우엔 불가피한 것일까요?
이에 대한 줄라이의 생각이 궁금합니다😊

yeahg-dev committed:
projects: JuiceMaker-MVVM 프로젝트 생성

yeahg-dev committed:
📱project: SPM으로 RxSwift 라이브러리 추가

yeahg-dev committed:
🔨feat: Fruit 데이터 타입 정의, FruitRepository 구현

yeahg-dev committed:
🔨feat: FruitJuice 타입 정의

yeahg-dev committed:
🔨feat: JuiceMaker 타입정의, fruitStockObservable구현

- ViewModel이 Repository의 과일 데이터 접근을 위한 API 제공위함

yeahg-dev committed:
🔨feat: JuiceMakerViewModel 정의, fruitStockObservable메서드 구현

- 과일재고를 방출하기 위해서 구현

yeahg-dev committed:
🔨feat: 쥬스 주문 화면, 과일 재고 수정 화면 UI 구현

yeahg-dev committed:
🔨feat: JuiceMaker화면 과일재고 로드 기능 구현

yeahg-dev committed:
🔨feat: 쥬스 주문 로직, alert제외 UI바인드 구현

- 재료가 모자람에도 쥬스가 주문되는 버그 있음(재료가 2개인 쥬스 주문시)

yeahg-dev committed:
🔨feat: FruitStock뷰에 기존과일재고 반영 기능 구현

yeahg-dev committed:
🗂file: 파일 구조 변

yeahg-dev committed:
🔨feat: LaunchScreen UI 구현

yeahg-dev committed:
🔨feat: FruitStock뷰에 네비게이션바 UI추가, done버튼 탭시 dismiss 구현

yeahg-dev committed:
📱project: RxCocoa 라이브러리 추가

yeahg-dev committed:
🔨feat: 과일재고 수정 기능 구현

yeahg-dev committed:
🪄fix: FruitStock뷰에서 stepper이벤트 바인드 오류 수정(오류 잔재)

-  transform이 호출된 후  FruitStockVC에서 loadFruitStock이 호출되기때문에, FruitStockVC생성 시엔 initialFruitStock이 nil임. 따라서 nil병합 연산자로 우선 10을 넣어서 일시적으로 오류 수정
- stepper를 통한 과일재고 수정후, 수정된 재고를 JuiceMaker로 부터 받아오는 로직 수정 필요:

yeahg-dev committed:
🪄fix: 재고 0이하로 수정할 때 앱 종료되는 오류 fix

- 재고 수정 요청에 대한 결과를 FruitStockModification타입으로 정의
- success, deficientStockFailure 각 케이스에 대해 뷰모델에서 뷰에 보여질 데이터 결정
- Repository에서 최신 데이터를 받아와서 뷰에 바인드하도록 로직 수정. 뷰모델과 모델간의 의존성 줄이기 위함

yeahg-dev committed:
🫧 style: 코드컨벤션 준수

yeahg-dev committed:
🔨fix: FruitStock뷰 딸기외 과일 transform 구현

yeahg-dev committed:
🎈chore: rxSwift 패키지 중복 추가 제거

yeahg-dev committed:
🫧 style: 코드컨벤션 준수

July911 commented on JuiceMaker-MVVM/JuiceMaker-MVVM/ViewModel/FruitStockViewModel.swift:
```diff
@@ -0,0 +1,231 @@
+    // MARK: - Property
+    
+    private let juiceMaker = JuiceMaker()
+    private let disposeBag = DisposeBag()
```
MVVM과 반응형 프로그래밍에서 얻는 장점은 뷰가 수동적이게 된다는 점인 것 같습니다. View까지 이어지는 스트림에서 중간에 스트림이 분기처리되면 훗날 프로젝트가 커질 때 스트림을 모두 이해하기 어려울 것 같습니다. 
DisposeBag 같은 경우에는 dispose되었을때만 필요한 것 같아서 ViewController을 제외하면 굳이 필요없을 것 같습니다 :)

yeahg-dev replied:
스트림이 분기되면 모든 스트림을 이해하기 어렵고, 또 예측하기 어렵다는 단점이 있겠네요!
뷰에서만 구독하는 방침으로 수정해보도록하겠습니다🔨

July911 commented on JuiceMaker-MVVM/JuiceMaker-MVVM/ViewModel/FruitStockViewModel.swift:
```diff
@@ -0,0 +1,231 @@
+        let pineappleStockObservable: PublishSubject<String?>
+        let watermelonStockObservable: PublishSubject<String?>
+        let bananaStockObservable: PublishSubject<String?>
+        let notificationObservable: PublishSubject<UserNotification>
```
ViewController는 ViewModel을 소유하고 있습니다 ! 
그래서 subject를 통해 onNext로 데이터를 전달해줘도 상관없을 수도 있다고 생각합니다. 하지만 통일성을 위하여 Observable로 통일 시켜주나 Driver를 사용한 레퍼런스들도 많이 있는 것 같습니다 :) 
참고해보시면 좋을 것 같아요 !

yeahg-dev replied:
뷰모델에서 인풋과 아웃풋의 타입을 통일시켜줘야한다는 의견으로 이해했는데 맞을까요??
그렇다면 통일성을 줘야하는 필요성에 대해서 좀 더 설명 부탁드려도 될까요?

다른 rx와 MVVM 레퍼런스들도 참고 해보도록하겠습니다🙂

July911 replied:
다른 레퍼런스들도 subject를 사용하는 경우도 종종 있습니다. 
단지 지금 상황에서는 subject일 필요가 있나 ...? 라는 생각이 있습니다 :) 
굳이 observable로 할 수 있는걸 subject로 하신 이유가 있으신가요 ??

yeahg-dev replied:
스테퍼 이벤트가 발생할 때마다 모델의 수정을 거쳐서 새로운 데이터를 output으로 흘려주어야합니다. observable은 새로운 observable을 생성하지 않는 한, 아이템을 방출만 할 수 있기 때문에 Subject로 정의해주었습니다.

July911 commented on JuiceMaker-MVVM/JuiceMaker-MVVM/ViewModel/FruitStockViewModel.swift:
```diff
@@ -0,0 +1,231 @@
+    func transform(input: Input) -> Output {
+
+        input.strawberryStepperValueObservable?
+            .map({[weak self] stepperValue in
```
캡쳐리스트에 weak self를 담아서 순환참조를 피하신 모습 너무 좋습니다.
Operator중에  `withUnretained`라는 것도 있는데 한번 참고해보시면 좋을 것 같습니다 !

yeahg-dev replied:
감사합니다!😃
공부해서 한번 적용해보았습니다~

July911 commented on JuiceMaker-MVVM/JuiceMaker-MVVM/ViewModel/FruitStockViewModel.swift:
```diff
@@ -0,0 +1,231 @@
+                (self?.initialStrawberryStock ?? 10) + Int(stepperValue)
+            })
+            .flatMap({ newValue in
+                self.juiceMaker.modifiedFruitStockObservable(of: .strawberry, with: newValue)
```
flatMap을 통해 Observable을 바꿔주신 부분 너무 좋습니다 ! 
한번 생각해보아요 -> 
Observable로 네트워킹을 하는 상황에서 이전 요청이 끝나지 않은 상태에서 새로운 요청이 들어오는 경우 어떻게 이전에 요청했던 task를 멈출 수 있을까요?

yeahg-dev replied:
Observable의 연산자 중 하나의 스트림만 흐르도록 하는 연산자를 찾아보았는데 찾지 못했습니다🥲

대신 태스크를 종료하는 방법을 찾아보던 중 `unsubscribe()`라는 메서드를 알게되었습니다. 
observable을 complete시키지 않으니, 새로운 네트 워크 요청을 트리거하는 곳에서 네트워크 통신을 하는 observable을 `unsubscribe` 해주는 방법으로 이전에 요청햇던 태스크를 멈출 수 있을 것 같습니다...!
https://stackoverflow.com/questions/45333939/angular-what-is-the-preferred-way-to-terminate-observables

위 방법으로 구현이 가능할까요...?🤔
제가 알아내지 못한 더 좋은 방법이 있다면 공유해주시면 감사하겠습니다!!

July911 replied:
flatMapLastest가 무엇이고 어떠한 동작을 할까요 ?

July911 commented on JuiceMaker-MVVM/JuiceMaker-MVVM/ViewModel/FruitStockViewModel.swift:
```diff
@@ -0,0 +1,231 @@
+            .flatMap({ newValue in
+                self.juiceMaker.modifiedFruitStockObservable(of: .strawberry, with: newValue)
+            })
+            .subscribe(
```
subscribe가 꼭 필요할까요 ? :) 
do 를 통해 Observable의 파생없이 처리를 해줄 수 있을 것 같아요 !

yeahg-dev replied:
`do`를 사용해서 subscribe하지 않고 새로운 이벤트 받아 를 Subject에 흘려보내는 로직으로 수정해주었습니다!

July911 commented on JuiceMaker-MVVM/JuiceMaker-MVVM/ViewModel/JuiceMakerViewModel.swift:
```diff
@@ -0,0 +1,71 @@
+    // MARK: - Property
+    
+    private let juiceMaker = JuiceMaker()
+    private let disposeBag = DisposeBag()
```
ViewModel에서 disposeBag을 제거해주세요 !

July911 commented on JuiceMaker-MVVM/JuiceMaker-MVVM/ViewModel/JuiceMakerViewModel.swift:
```diff
@@ -0,0 +1,71 @@
+    // MARK: - Input/Output
+    
+    struct Input {
+        let juiceOrder: PublishSubject<FruitJuice>
```
ViewController가 ViewModel의 인터페이스를 통해서 본인의 UIComponents에 바인딩을 해주는데
1.ViewModel은 ViewController를 알고 있을까요 ? 
알고 있다면 굳이 UIkit을 import안할 이유가 없을 것 같습니다.
2. 모른다면 지금 구조에서는 ViewController에 ViewModel이 영향을 줄 수 있는 구조인 것 같습니다.
훗날 클린아키텍쳐로 넘어갔을 때 독립된 DomainLayer를 유지하기 위해서 Subject를 프로퍼티로 가지지 않는 상황과 비슷한 상황으로 보입니다 ! 수정부탁드려요 :)

July911 commented on JuiceMaker-MVVM/JuiceMaker-MVVM/ViewModel/JuiceMakerViewModel.swift:
```diff
@@ -0,0 +1,71 @@
+    
+    struct Output {
+        let orderSuccess: BehaviorSubject<Bool>
+        let alertMessage: BehaviorSubject<String>
```
PublishSubject와 BehaviorSubject의 차이와 장단점에 대하여 적어주세요 !

yeahg-dev replied:
Publish Subject는 원본 옵저버블을 구독한 시점 후에 방출되는 아이템을 방출하고, Behavior Subject는 원본 옵저버블이 구독 시점 전 가장 최근에 방출한 아이템 또는 방출한 것이 없다면 기본값을 먼저 방출하기 시작합니다. 
그래서 PublishSubject는 과거에 갱신된 데이터는 무시하고 구독된 시점부터의 데이터를 받고 싶을때,  Behavior Subject는 구독하는 시점에 과거에 갱신된 데이터 중 가장 최근의 데이터를 활용하고 싶을 때 사용할 수 있습니다.

July911 commented on JuiceMaker-MVVM/JuiceMaker-MVVM/View/FruitStockViewController.swift:
```diff
@@ -0,0 +1,91 @@
+            .subscribe(onNext: {[weak self] stock in
+                self?.strawberryStockLabel?.text = stock
+            })
+            .disposed(by: disposeBag)
```
subscribe로 바인딩을 진행해준다면 굳이 RxCocoa를 import하지 않아도 될 것 같습니다. 
Subscribe vs Bind vs  Driver  의 차이를 설명하고 뷰 바인딩 상황에 특화된 친구가 있는지 알아보아요 :)

yeahg-dev replied:
`subscribe` `bind` `driver` 3 가지 모두 옵저버블을 구독하는 방법이고, 차이점으로는 어떠한 이벤트를 핸들링할 수 있는지와 어디서 observinge되는지가 다릅니다. 
`subscribe`는  next, error, complete 각 이벤트에 대한 핸들러를 구현할 수 있습니다. 
`bind`는 next이벤트에 대한 핸들러만 구현할 수 있습니다.
`driver`는 mainscheduler에서 observing하도록 하며, next, complete만 핸들링 할 수 있습니다. 

뷰 바인딩에는 메인스레드에서 동작하도록 구현된 `driver`가 편리하겠네요😊

July911 commented on JuiceMaker-MVVM/JuiceMaker-MVVM/View/FruitStockViewController.swift:
```diff
@@ -0,0 +1,91 @@
+    
+    // MARK: - bind
+    
+    func bindUI() {
```
Public 메서드일 이유가 있을까요 ?

yeahg-dev replied:
private으로 접근제어 수정했습니다!

July911 commented on JuiceMaker-MVVM/JuiceMaker-MVVM/View/FruitStockViewController.swift:
```diff
@@ -0,0 +1,91 @@
+import RxSwift
+import RxCocoa
+
+class FruitStockViewController: UIViewController {
```
상속의 여지가 없다면 final 키워드를 통해서 성능을 향상시킬 수 있을 것 같습니다 !

July911 commented on JuiceMaker-MVVM/JuiceMaker-MVVM/View/JuiceMakerViewController.swift:
```diff
@@ -0,0 +1,112 @@
+        }
+        
+        juiceMakerViewModel.fruitStockObservable(of: fruit)
+            .map{String($0)}
```
수정하셔도 되고 안하셔도 되는 코멘트입니다. 
ViewModel은 ViewController에 표시 될 정보를 가공해서 줍니다. 
여기서 데이터를 변경해 줄 이유가 있을까요 ? 변경해서 넘겨주는 방향은 어떻게 생각하시나요 ?

yeahg-dev replied:
변경해서 넘겨주는 방향이 더 좋은 것 같아요! 뷰는 최대한 멍청하게!!
수정했습니다😁

July911 commented on JuiceMaker-MVVM/JuiceMaker-MVVM/View/JuiceMakerViewController.swift:
```diff
@@ -0,0 +1,112 @@
+        self.input.juiceOrder.onNext(StrawberryPeachJuice())
+    }
+    
+    @IBAction func orderStrawberryJuice(_ sender: UIButton) {
```
훗날 버튼이나 TableView, CollectionView의 cellTap 상황에서는 rx. 으로 시작하는 많은 Reactive Extension이 존재합니다. 
코드의 중복을 막아줄 수도 있고 더 쉽게 이해되기도 합니다 ! :) 
한번 알아보시면 좋을 것 같아요 !

yeahg-dev replied:
rxCocoa로 버튼 탭에 대한 옵져버블을 생성하는 방향으로 수정해보았습니다!
rx스러운 코드로 통일성이 생긴 것 같습니당🤓
```swift
 self.strawberryJuiceButton?.rx.tap
            .bind {
                juiceOrder.onNext(StrawberryJuice())
            }
            .disposed(by: disposeBag)
```

July911 commented on JuiceMaker-MVVM/JuiceMaker-MVVM/Repository/FruitRepository.swift:
```diff
@@ -0,0 +1,67 @@
+    
+    static let shared = FruitRepository(defaultFruitStock: 10)
+    
+    var storage: [Fruit: Int] = [:]
```
객체지향의 사실과 오해에서 객체와 객체는 인터페이스 ( 메서드 ) 로 소통한다고 배운 것 같아요 :) 
프로퍼티가 외부에서 쓰일 일이 있을까요 ?

yeahg-dev replied:
private으로 객체 내부에서만 수정이 가능하도록 하는게 좋을 것 같네요!

July911 commented on JuiceMaker-MVVM/JuiceMaker-MVVM/Repository/FruitRepository.swift:
```diff
@@ -0,0 +1,67 @@
+        return observable
+    }
+    
+    func update(_ fruit: Fruit, with newValue: Int) -> Observable<FruitRepository.Result> {
```
Observable을 생성하는 just와 closure를 열어서 생성하는 방법 모두 잘 사용하신 것 같아요 :) 
빠른 시간안에 많이 성장하신 것 같습니다 대단하세요 !

yeahg-dev replied:
감사합니다😆

July911 commented on JuiceMaker-MVVM/JuiceMaker-MVVM/Repository/FruitRepository.swift:
```diff
@@ -0,0 +1,67 @@
+    
+    func update(_ fruit: Fruit, with newValue: Int) -> Observable<FruitRepository.Result> {
+        Observable.create { observer in
+            self.storage.updateValue(newValue, forKey: fruit)
```
* 진짜 정말정말 개인적인 의견입니다 !  
많은 레퍼런스들을 보면 Observable을 생성할 때 emitter라는 이름으로 많이 쓰는 것 같습니다. 아마도 추측하기엔 곰튀김님 영상에서 이러한 네이밍이 나와서가 아닌가 ...? 라는 생각을 해봅니다 ! 
굳이 안바꾸고 emitter라는 네이밍을 굳이 쓰는 이유는 이미 많이 그렇게 사용하고 눈에 익어서라고 생각합니다. 아는 내용이 나오면 눈에 익숙해서 보기 편하니까요 !

yeahg-dev replied:
참고하겠습니다~!

July911 commented on JuiceMaker-MVVM/JuiceMaker-MVVM/App/SceneDelegate.swift:
```diff
@@ -0,0 +1,52 @@
+        // If using a storyboard, the `window` property will automatically be initialized and attached to the scene.
+        // This delegate does not imply the connecting scene or session are new (see `application:configurationForConnectingSceneSession` instead).
+        guard let _ = (scene as? UIWindowScene) else { return }
+    }
```
나중에 Coordinator패턴을 사용하시게 된다면 AppCoordinator는 여기에 위치 할 것 같습니다 :) 참고하셔용

yeahg-dev replied:
언능 다음 프로젝트에서 써보고 싶네요🚀

July911 commented on JuiceMaker-MVVM/JuiceMaker-MVVM/View/FruitStockViewController.swift:
```diff
@@ -0,0 +1,91 @@
+    
+    private lazy var input = FruitStockViewModel.Input(
+        strawberryStepperValueObservable: self.strawberryStepper?.rx.value.asObservable(),
+        peachStepperValueObservable: self.peachStepper?.rx.value.asObservable(),
```
asObservable 로 Observable로 바꿔주신 모습  정말 좋아요 :)

July911 commented on JuiceMaker-MVVM/JuiceMaker-MVVM/View/FruitStockViewController.swift:
```diff
@@ -0,0 +1,91 @@
+        watermelonStepperValueObservable: self.watermelonStepper?.rx.value.asObservable(),
+        bananaStepperValueObservable: self.bananaStepper?.rx.value.asObservable())
+    
+    private lazy var output = fruitStockViewModel.transform(input: input)
```
디스코드로 이야기를 나눴을 때 이러한 레퍼런스를 보시고 하셨다고 말씀해주셨는데 어떤 레퍼런스인지 공유해주세요 :) 
그리고 지연할당까지 하면서 함수를 분리해줘야 하나 ..? 라는 생각도 듭니다 ! 
어떻게 생각하세요 ?

yeahg-dev replied:
https://blog.gangnamunni.com/post/HealingPaperTV-ViewModel-Test/
위 블로그를 참고하였습니다. 
그런데 input과 output은 `bindUI()` 메서드 내부에서만 사용하니 굳이 프로퍼티로 선언해주지 않아도 될것 같네요!

July911 commented on JuiceMaker-MVVM/JuiceMaker-MVVM/Model/JuiceMaker.swift:
```diff
@@ -0,0 +1,128 @@
+                guard let juiceToMake = juice else {
+                   return
+                }
+                self.consumeIngredients(of: juiceToMake)
```
정말정말정말 잘하셨네요. 
map을 통해 데이터 변형을 해주고 take를 통해서 횟수만큼 일처리가 되게 해주신 모습 정말 훌륭합니다. 
scan을 통해서 이전 데이터를 합해서 넣어주시는 것도 좋고 만약 매번 그렇게 할 이유가 없다면 startWith를 통해서 초기값을 주는 경우도 있습니다 ! 
잘해주셨어요 !!!!!! 릴리 정말 빨리 잘해지시는군요

yeahg-dev replied:
오호~ 그런 방법도 있겠군요!!

July911 commented on JuiceMaker-MVVM/JuiceMaker-MVVM/Model/JuiceMaker.swift:
```diff
@@ -0,0 +1,128 @@
+    }
+    
+    private func hasSufficentIngredients(of juice: FruitJuice) -> BehaviorSubject<Bool> {
+        // 쥬스 재료로 사용될 과일의 가용여부 담을 subject
```
subject일 필요가 있는지 한번만 더 생각해보아요 :)

yeahg-dev replied:
observable을 리턴하도록 수정했습니다:)

July911 commented on JuiceMaker-MVVM/JuiceMaker-MVVM/Model/JuiceMaker.swift:
```diff
@@ -0,0 +1,128 @@
+                                    observer.onNext(FruitStockModification.success)}
+                            })
+                            .disposed(by: disposeBag)
+                    } else if result.element == false {
```
구독안에 구독이라 .. 어려운 구조네요 
마치 탈출클로저를 통해 비동기작업을 해주었던 무한 콜백이 생각나요 
구조를 한번 바꿔보시는 것도 좋을 것 같아요 :) 
그리고 do 를 사용하여 구독을 후로 미루는 상황이 온다면 언젠가 한번은 ( 아마도 ViewController겠죠? ) 구독을 꼭 해주어야 스트림이 시작된다는 사실 ! 
정말정말 잊으시면 안됩니다 ....

yeahg-dev replied:
네! 명심 또 명심 할게욧🫡

July911 commented on JuiceMaker-MVVM/JuiceMaker-MVVM/Model/JuiceMaker.swift:
```diff
@@ -0,0 +1,128 @@
+                                    observer.onNext(FruitStockModification.success)}
+                            })
+                            .disposed(by: disposeBag)
+                    } else if result.element == false {
```
구독안에 구독이라 .. 어려운 구조네요 
마치 탈출클로저를 통해 비동기작업을 해주었던 무한 콜백이 생각나요 
구조를 한번 바꿔보시는 것도 좋을 것 같아요 :) 
그리고 do 를 사용하여 구독을 후로 미루는 상황이 온다면 언젠가 한번은 ( 아마도 ViewController겠죠? ) 구독을 꼭 해주어야 스트림이 시작된다는 사실 ! 
정말정말 잊으시면 안됩니다 ....

yeahg-dev replied:
네! 좋은 지적 감사합니다👍
수정해보았습니다

July911 commented on JuiceMaker-MVVM/JuiceMaker-MVVM/Model/JuiceMaker.swift:
```diff
@@ -0,0 +1,128 @@
+    ) -> Observable<FruitStockModification> {
+        Observable<FruitStockModification>.create { observer in
+            self.isAbleToModifyFruitStock(of: fruit, with: newValue)
+                .subscribe { result in
```
폴더명을 Model로 지으셨네요 :) 
MVVM-C 클린아키텍쳐를 적용한다면 이 부분은 UseCase가 되겠죠? 
UseCase에 disposeBag이 필요할까요 ? 
구독이 필요할까요 ..?

yeahg-dev replied:
구독을 하지 않고 do를 사용해 개선해보았습니다!

July911 commented on JuiceMaker-MVVM/JuiceMaker-MVVM/View/JuiceMakerViewController.swift:
```diff
@@ -0,0 +1,112 @@
+import UIKit
+import RxSwift
+
+class JuiceMakerViewController: UIViewController {
```
보통은 ! 파일을 나눌때 각 Scene별로 ViewController ViewModel Coordinator 를 한 폴더안에 넣어주는 것 같습니다 ! 참고하시면 좋을 것 같아요 !

yeahg-dev replied:
오홋! 🤩 감사합니다:)

July911 commented on JuiceMaker-MVVM/JuiceMaker-MVVM/View/JuiceMakerViewController.swift:
```diff
@@ -0,0 +1,112 @@
+import UIKit
+import RxSwift
+
+class JuiceMakerViewController: UIViewController {
```
상속의 여지가 없다면 final키워드를 추가해주세요 !

yeahg-dev replied:
추가해주었습니다!👍

July911 reviewed (commented):
정말 잘해주셨습니다 :) 
역시 잘하셔서 금방 RxSwift도 이해하시는 것 같아요 ... 저는 처음할 때 훨씬 못했습니다 ! 
질문주신 부분은 점심먹고 마저 적겠습니다 ! 
코멘트 남겨주시고 수정해주세요 :) ( 릴리가 생각하셨을때 수정이 필요없겠다 싶은 부분은 안하셔도 됩니다 ! 코드엔 정답은 없으니까요 ! )

yeahg-dev committed:
✨refactor: FruitStockViewModel transform 로직 수정

- 뷰모델에서 옵저버블 구독하지 않도록 수정

yeahg-dev committed:
🩹fix: 컴파일 타임 오류 수정

yeahg-dev committed:
✨refactor: [weak self] 대신 withUnretained로 수정

- 가독성 향상을 위함

yeahg-dev committed:
✨refactor: 접근제어자 수정

- 프로퍼티를 객체 내부에서만 변경 가능하도록 하기 위함

yeahg-dev committed:
✨refactor: final키워드 추가하여 성능최적화

yeahg-dev committed:
✨refactor: FruitStockViewModel의 프로퍼티 제거, transform 메서드 리팩터링

- Repository로부터의 데이터 스트림을 끊지 않기위해서 take와 combineLatest 활용
- updatedFruitStock(of:, input:)을 만들어서 중복코드 제거

yeahg-dev committed:
✨refactor: FruitStockVC bindUI() 접근제어레벨 수정

yeahg-dev committed:
✨refactor: FruitStockVC의 input, output 지연저장프로퍼티를 bindUI()로 이동

yeahg-dev committed:
✨refactor: 주스주문 이벤트를 IBAction에서 rxCocoa로 수정

yeahg-dev committed:
✨refactor: JuiceMakerVC에서 옵져버블의 타입변환 과정 제거

yeahg-dev committed:
✨refactor: JuiceMakerViewModel Output 모델링 수정

- 각 과일 재고에 대한 옵져버블을 Output으로 정의하여 viewWillAppear와 주스버튼 탭이 발생할 때마다 업데이트 되도록 변경
- 뷰에서 사용되는 과일재고를 뷰모델로 직접적으로 정의하여 가독성을 높임

yeahg-dev committed:
✨refactor: flatMap으로 JuiceMaker에서 구독을 제거

yeahg-dev committed:
✨refactor: 메서드명 수정

yeahg-dev committed:
🩹fix: hasSufficentIngredient 구현 수정

- 기존 과일 재고를 확인하는 로직엔 버그 존재
- zip 을 사용해서 각 과일 재고 가능 여부를 새로운 bool값으로 리턴하도록함

yeahg-dev committed:
✨refactor: 각 주문 버튼을 Input으로 수정, 재고부족 에러 핸들링 추가

- 각 주문 버튼을 옵저버블 Input으로 정의하여 뷰 레이어에서의 비지니스로직 제거
- 버튼 옵저버블을 merge 연산자로 구독하도록 구현
- 각 버튼 Input에 대한 동일한 형식의 Output을 Publish로 구현
- 재고 부족에 대한 에러를 Usecase에서 정의 후 ViewModel에서 에러핸들링
- 에러 방출시 스트림 종료를 방지하기 위해 retry 구현

yeahg-dev committed:
✨refactor: 재고수정 에러핸들링 구현 수정

- Usecase에서 재고 수정이 불가하다고 판단될 시 에러를 방출하도록 설계
- 뷰모델에서 에러핸들링 구현(alertMessage에 에러메시지 전달)
- 에러 발생하면 스트림이 종료되기 때문에 retry(when:) 사용하여, stepperValue발생할 때 마다 재구독 하도록 구현

yeahg-dev committed:
🫧 style: debug() 삭제

yeahg-dev committed:
🫧 style: 불필요한 코드 제거, 메서드명 수정, 메서드 위치 수정

yeahg-dev committed:
✨refactor: JuiceMaker에서 disposebag 제거

yeahg-dev committed:
🗂file: 폴더 구조 변경

yeahg-dev committed:
✨refactor: 주문버튼 탭시 최신 과일재고 반영하도록 구현

yeahg-dev committed:
🔨feat: JuiceMakerVC presentAlert 구현

yeahg-dev committed:
🔨feat: FruitStockVC presentAlert 구현

yeahg-dev committed:
✨refactor: JuiceMakerNotification으로 alertMessage 래핑

yeahg-dev commented:
안녕하세요 줄라이! 
꼼꼼하게 리뷰해주시고 추가 키워드들을 알려주셔서 리팩터링을 진행하며 많은 공부가 되었습니다.
감사합니다!!!!😊

리뷰주신 부분들과 이전에 구현하지 못했거나 오류가 있었던 부분 추가적으로 수정했습니다.
수정 내용 정리해보자면 아래와 같습니다:)

### Input, Output 패턴 적용
- 기존에 ViewController에서 프로퍼티로 가지던  뷰모델의 `input`, `output`을 `transform` 메서드 내부에서 생성하도록 수정했습니다. 
-  기존 과일주스 주문 버튼 탭에 대한 하나의 input을 각 과일 주스 버튼 탭에 대한  input으로 만들어 주었습니다.  

### Usecase에서 disposeBag제거
- Usecase에서 `Subject` 와 `disposeBag`을 제거하여 스트림이 끊기지 않도록 개선했습니다.

### Usecase서 에러 방출과 `retry`를 사용한 에러핸들링
- 쥬스 주문시 과일 재고 부족, 0이하로 과일재고를 수정 하려는 시도 : 2가지 에러 상황을 Usecase에서 방출, 에러를 뷰모델에서 핸들링하도록 구현했습니다. 뷰모델에서 `do(onError:)`로 에러에 대한 알럿 이벤트를 흘려보내어 사용자에게 알럿을 전달합니다. 
- `error`가 방출된 스트림은 종료되기 때문에 `retry` 와 `retry(when:)`을 호출하여 재구독을 해주었습니다. 

### UI추가 구현
- UIAlertController를 활용, 이벤트 결과에 대한 알림을 사용자에게 표시 기능을 구현했습니다. 


## 고민한 점
### 뷰와 최신 데이터의 동기화
기존 구현에서는 과일 재고가 수정되어야 할 때마다 뷰모델에서 `fruitObservable`을 bind(구독)해서 output(PublishSubject)으로 흘려보내도록 구현했습니다. 
피드백 주신 것과 같이 뷰모델에서 disposeBag을 제거하고, 일관적인 스트림을 지키기 위해  repository에 과일 재고 Subject을 생성해 해결하고자 시도했었습니다. 

그런데 주문 버튼 탭 이벤트가 들어올 때마다 비정상적으로 과일재고가 감소되는 오류가 있었고, 디버깅을 시도해봤지만 원인을 찾지 못했습니다.
그래서 disposeBag을 제거하지 못하고, 과일 재고가 수정되어야하는 input이 들어올 때마다 `fruitObservable`을 bind(구독)해서 output(PublishSubject)으로 흘려보내도록 수정했습니다. 

아직 제가 경험과 지식이 부족해서, 이와 같이 구현한것이 최선이네요ㅜㅜ 
이러한 사례에 대한 레퍼런스를 좀 더 찾아보도록 할게요🔥
```swift
 let buttonSubscribe = Observable.merge(input.viewWillAppear, strawberryAction!, peachAction!, strawberryPeachAction!, pineappleAction!, watermelonAction!, watermelonPineappleAction!, bananaAction!)
                    .do(onNext: { self.fruitStock(of: .strawberry)
                            .bind{ stock in
                                strawberryStock.onNext(stock)
                            }.dispose()
                        self.fruitStock(of: .peach)
                            .bind{ stock in
                                peachStock.onNext(stock)
                            }.dispose()
                        self.fruitStock(of: .pineapple)
                            .bind{ stock in
                                pineappleStock.onNext(stock)
                            }.dispose()
                        self.fruitStock(of: .watermelon)
                            .bind{ stock in
                                watermelonStock.onNext(stock)
                            }.dispose()
                        self.fruitStock(of: .banana)
                            .bind{ stock in
                                bananaStock.onNext(stock)
                            }.dispose()})
        .retry()
```

July911 commented on JuiceMaker-MVVM/JuiceMaker-MVVM/Presenter/FruitStockModificationScene/FruitStockViewController.swift:
```diff
@@ -0,0 +1,109 @@
+    // MARK: - bind
+    
+    private func bindUI() {
+        let input = FruitStockViewModel.Input(
```
지연할당을 시켜주지 않고 함수 내부에서 처리하신 모습 정말 훌륭합니다 :)

July911 commented on JuiceMaker-MVVM/JuiceMaker-MVVM/Presenter/FruitStockModificationScene/FruitStockViewController.swift:
```diff
@@ -0,0 +1,109 @@
+        let output = fruitStockViewModel.transform(input: input)
+        
+        output.notificationObservable?
+            .subscribe { userNotification in
```
지금도 정말 좋은 방법이긴 하나, Reactive Extension으로 rx.으로 연결되는 메서드를 만들 수 있을 것 같습니다. :) 
저는 
https://github.com/July911/ChildOfWeather/blob/step4/ChildOfWeather/Presentation/Extension/RxCocoa%20%2B%20Extension.swift
이러한 방법으로 Extension을 만들어서 관리해주었습니다 :)

yeahg-dev replied:
오호~ 익스텐션을 활용하면 가독성이 더 좋아지겠네요!
공부해보도록할게요! 감사합니다😊

July911 commented on JuiceMaker-MVVM/JuiceMaker-MVVM/Presenter/FruitStockModificationScene/FruitStockViewController.swift:
```diff
@@ -0,0 +1,109 @@
+            }.disposed(by: disposeBag)
+        
+        output.strawberryStockObservable?
+            .withUnretained(self)
```
지난번 코멘트때 언급드렸던 withUnretained(self를 사용해보셨군요 !!! 
정말 오퍼레이터에 대한 이해와 습득 속도가 빠르신 것 같습니다 :) 
이번에는 저 오퍼레이터 사용의 주의할 점에 대하여 알아볼까요 ? https://minios.tistory.com/46

yeahg-dev replied:
`error`와 `completed` 이벤트에선 `[weak self]`를 또 다시 작성해줘야하는 불편한 점과,
여러번 사용시 코드가 길어지고 오히려 가독성이 떨어질 수도 있겠군요!

July911 commented on JuiceMaker-MVVM/JuiceMaker-MVVM/Presenter/FruitStockModificationScene/FruitStockViewModel.swift:
```diff
@@ -0,0 +1,109 @@
+        let peachStepperValueObservable: Observable<Double>?
+        let pineappeldStepperValueObservable: Observable<Double>?
+        let watermelonStepperValueObservable: Observable<Double>?
+        let bananaStepperValueObservable: Observable<Double>?
```
옵셔널로 각 프로퍼티를 정의해주신 이유가 있나요 ?

yeahg-dev replied:
스트림이 시작되는 `UIButton` IBOutlet 프로퍼티들을 옵셔널로 정의해서, 옵셔널 체이닝으로 Observable도 옵셔널이 되었습니다.

nil 병합 연산자로 옵셔널 바인딩을 해서 프로퍼티를 논옵셔널 타입으로 정의할 수도 있을 것 같습니다.
그런데 이렇게 구현할 경우, 만약 실제 IBOutlet이 nil인 경우 = 화면에 없는 경우에 깡통 이벤트를 전달해주는 건 의도하지 않은 실행흐름을 만들 수 있다고 생각하기에... 옵셔널로 두는게 더 나을 수 있겠다라고 생각했습니다.

줄라이는 어떻게 생각하시나요??

July911 commented on JuiceMaker-MVVM/JuiceMaker-MVVM/Presenter/FruitStockModificationScene/FruitStockViewModel.swift:
```diff
@@ -0,0 +1,109 @@
+            .flatMap {
+            self.juiceMaker.modifyFruitStock(of: fruit, with: $0)
+            }
+            .share(replay: 1)
```
진짜 ... 대단하신 것 같아요 
너무 훌륭하게 메서드 설계를 해주신 것 같습니다.

July911 commented on JuiceMaker-MVVM/JuiceMaker-MVVM/Presenter/FruitStockModificationScene/FruitStockViewModel.swift:
```diff
@@ -0,0 +1,109 @@
+                self.juiceMaker.fruitStockObservable(of: fruit)
+            }
+            .map{String($0)}
+            .retry(when: { _ in
```
retry , share .... 멋져요 ...

July911 commented on JuiceMaker-MVVM/JuiceMaker-MVVM/Presenter/JuiceMakerScene/JuiceMakerViewModel.swift:
```diff
@@ -0,0 +1,242 @@
+        let pineappleStock = PublishSubject<String>()
+        let watermelonStock = PublishSubject<String>()
+        let bananaStock = PublishSubject<String>()
+        
```
2가지 방법으로 생각해 볼 수 있을 것 같습니다. 
릴리의 방법이 틀린게 아니라 repo에 subject를 가지고 있는 
https://github.com/July911/iOS-JuiceMakerRxSwift 
이러한 방법
그리고 이러한 방법 둘 중 사용하는 곳에서 맞게 사용해주시면 좋을 것 같습니다 :)

yeahg-dev replied:
좋은 레퍼런스 제작해주셔서 감사합니다👍
다음 프로젝트 때 참고해보겠습니다~

July911 commented on JuiceMaker-MVVM/JuiceMaker-MVVM/Presenter/JuiceMakerScene/JuiceMakerViewModel.swift:
```diff
@@ -0,0 +1,242 @@
+        let pineappleStock = PublishSubject<String>()
+        let watermelonStock = PublishSubject<String>()
+        let bananaStock = PublishSubject<String>()
+        
```
2가지 방법으로 생각해 볼 수 있을 것 같습니다. 
릴리의 방법이 틀린게 아니라 repo에 subject를 가지고 있는 
https://github.com/July911/iOS-JuiceMakerRxSwift 
이러한 방법
그리고 이러한 방법 둘 중 사용하는 곳에서 맞게 사용해주시면 좋을 것 같습니다 :)

July911 commented on JuiceMaker-MVVM/JuiceMaker-MVVM/Presenter/JuiceMakerScene/JuiceMakerViewModel.swift:
```diff
@@ -0,0 +1,242 @@
+                        self.fruitStock(of: .pineapple)
+                            .bind{ stock in
+                                pineappleStock.onNext(stock)
+                            }.dispose()
```
이렇게 사용하신 이유에 대해서 코멘트 한번 남겨주세요 :)

yeahg-dev replied:
`viewWillAppear`나 `buttonTapped`이벤트가 들어올 때마다 레포지토리로부터 과일의 재고를 읽어와서 뷰에 바인드해야합니다. Subject 대신 Observable을 리턴하면 처음 구독한 이후엔 새로운 재고를 뷰에  바인드할 수 없기 때문에 Subject로 구현을 했었습니다.

yeahg-dev replied:
음 그런데!! 레퍼런스 남겨주신 파일을 보고 아이디어가 떠올랐어요!
`viewWillAppear`와 `buttonTapped`들을 `Merge`한 다음 flatMap을 이용하여 새로운 observable을 리턴하도록 구현하면 bind를 사용하지 않고도 재고를 넘겨줄 수 있을 것 같네요!

아래와 같이 수정해보았으니 확인 한번만 부탁드려요🙌
수정 커밋👉 73d65c75b0a70bed00c5ee831bec75357ea0fb9d

July911 commented on JuiceMaker-MVVM/JuiceMaker-MVVM/Presenter/JuiceMakerScene/JuiceMakerViewModel.swift:
```diff
@@ -0,0 +1,242 @@
+                        self.fruitStock(of: .pineapple)
+                            .bind{ stock in
+                                pineappleStock.onNext(stock)
+                            }.dispose()
```
이렇게 사용하신 이유에 대해서 코멘트 한번 남겨주세요 :)

July911 commented on JuiceMaker-MVVM/JuiceMaker-MVVM/Presenter/JuiceMakerScene/JuiceMakerViewModel.swift:
```diff
@@ -0,0 +1,242 @@
+                                watermelonStock.onNext(stock)
+                            }.dispose()
+                        self.fruitStock(of: .banana)
+                            .bind{ stock in
```
bind는 drive , subscribe와 같은 맥락의 메서드가 아닌가요 ..?

yeahg-dev replied:
맞습니다..! 위 코드에서 구독을 없애보았습니다

July911 commented on JuiceMaker-MVVM/JuiceMaker-MVVM/Presenter/JuiceMakerScene/JuiceMakerViewModel.swift:
```diff
@@ -0,0 +1,242 @@
+                                watermelonStock.onNext(stock)
+                            }.dispose()
+                        self.fruitStock(of: .banana)
+                            .bind{ stock in
```
bind는 drive , subscribe와 같은 맥락의 메서드가 아닌가요 ..?

July911 commented on JuiceMaker-MVVM/JuiceMaker-MVVM/Presenter/JuiceMakerScene/JuiceMakerViewModel.swift:
```diff
@@ -0,0 +1,242 @@
+    // MARK: - Input/Output
+    
+    struct Input {
+        let viewWillAppear: Observable<Void>
```
정말 이러한 프로퍼티가 메서드가 쓰이고 있나요 ?

July911 replied:
정말 이러한 프로퍼티가 메서드가 쓰이고 있나요 ?

July911 commented on JuiceMaker-MVVM/JuiceMaker-MVVM/Usecase/JuiceMaker.swift:
```diff
@@ -0,0 +1,131 @@
+            let fruitAvailable = self.fruitRepository.read(fruit)
+                .flatMap { currentStock in
+                    currentStock >= ingredient.value ? Observable<Bool>.just(true) : Observable<Bool>.just(false)
+                }
```
Rx를 자유롭게 사용하시네요 ..
정말 2주간 많이 성장하신 것 같습니다 !

July911 replied:
Rx를 자유롭게 사용하시네요 ..
정말 2주간 많이 성장하신 것 같습니다 !

July911 commented on JuiceMaker-MVVM/JuiceMaker-MVVM/Repository/FruitRepository.swift:
```diff
@@ -0,0 +1,55 @@
+    
+    // MARK: - Method
+    
+    init(defaultFruitStock: Int) {
```
싱글톤은 private init 아니였나요 ...?

yeahg-dev replied:
깜빡했네요ㅎㅎ 수정했습니다!

July911 commented on JuiceMaker-MVVM/JuiceMaker-MVVM/Repository/FruitRepository.swift:
```diff
@@ -0,0 +1,55 @@
+    
+    // MARK: - Method
+    
+    init(defaultFruitStock: Int) {
```
싱글톤은 private init 아니였나요 ...?

July911 reviewed (commented):
정말 많이 성장하신게 눈에 보입니다. 

> Usecase서 에러 방출과 retry를 사용한 에러핸들링
쥬스 주문시 과일 재고 부족, 0이하로 과일재고를 수정 하려는 시도 : 2가지 에러 상황을 Usecase에서 방출, 에러를 뷰모델에서 핸들링하도록 구현했습니다. 뷰모델에서 do(onError:)로 에러에 대한 알럿 이벤트를 흘려보내어 사용자에게 알럿을 전달합니다.
error가 방출된 스트림은 종료되기 때문에 retry 와 retry(when:)을 호출하여 재구독을 해주었습니다.

이 부분은 정말 ... 정말 정말 잘하신 것 같습니다. 
error 상황을 정의하고 스트림이 끊어질 것 까지 생각하신 정도면 하산하셔도 될 것 같습니다.
publishSubject가 ViewModel에 있는 방법 외의 방법에 대해 많이 고민하신 것 같습니다.
참고 : https://github.com/July911/iOS-JuiceMakerRxSwift/blob/main/iOS-JuiceMakerRxSwift/FruitStroage.swift 
보시고 참고하시면 좋을 것 같습니다. 진작 이러한 코드들을 보여주며 설명 해드렸어야 되는데 
추상적인 말로만 말씀드려 혼란이 생긴 것 같습니다 ㅠㅠ 
다음 깃헙 프로젝트는 Coordinator가 나와 더 .. 힘들 수도 있지만 릴리의 성장 속도라면 분명 잘 하시리라 믿습니다. 화이팅입니다 :)

yeahg-dev committed:
✨refactor: JuiceMakerViewModel에서 disposeBag 제거

- 과일 재고가 수정되어야하는 input 이벤트들을 merge시켜 각 이벤트가 발생할 때마다 새로운 Observable을 리턴하도록 수정

yeahg-dev committed:
🩹fix: JuiceMaekr에서 에러리턴시 스트림이 종료되는 문제 해결

- retry연산자로 에러가 방출되면 재구독하도록 수정

yeahg-dev committed:
✨refactor: FruitStockVC가 dismiss시 JuiceMakerVC 재고업데이트 구현 시도

yeahg-dev committed:
✨refactor: 싱글톤 이니셜라이져 private 추가

yeahg-dev commented:
줄라이 항상 감사드립니다😊
이제 정말 마지막 리퀘스트가 될 것 같네요!

질문 남겨주신 부분에 답변 달았고,
JuiceMaker ViewModel에서 disposeBag을 없애는 로직으로 리팩터링 해보았습니다.

마지막으로 한가지 문제가 남아있는데, 한 번 검토 해주시면 감사하겠습니다.

### 문제 상황
FruitStockVC에서 재고를 수정한 뒤 JuiceMakerVC로 돌아올 때 수정된 재고를 반영하지 못하는 문제가 남아있었습니다.

### 해결 시도 
JuiceMakerVC의 뷰라이프 사이클 메서드로 input을 정의해서 수정된 재고를 반영하는 로직을 구현하고자 했지만,
현재 화면전환 방식은 present를 하더라도 JuiceMakerVC의 뷰라이프사이클에 변화가 없음을 확인했습니다. 

그래서 FruitStockVC에서 `Done`버튼이 탭될 때 JuiceMakerVC에서 호출되는 메서드를 input으로 사용해, 새로운 재고 옵저버블을 보내주도록 구현하고자 했습니다. 

```swift
class JuiceMakerViewController: UIViewController {

  private func bindUI() {
        
        let input = JuiceMakerViewModel.Input(
            viewWillAppear: self.rx.methodInvoked(#selector(UIViewController.viewWillAppear(_:))).map{_ in},
           // dismiss되는 시점에 호출될 메서드를 input으로 사용
            presentedViewDismissed: 
       self.rx.methodInvoked(#selector(JuiceMakerViewController.presentedViewControllerDidDismissed)).map{_ in}.debug(),
.... 
   }

  @objc private func presentedViewControllerDidDismissed() {
        return
    }
}

protocol FruitStockViewControllerDelegate {
    
    func didDismiss(_: UIViewController)
}

extension JuiceMakerViewController: FruitStockViewControllerDelegate {
    
    
    func didDismiss(_: UIViewController) {
        self.presentedViewControllerDidDismissed()
    }
}

class FruitStockViewController: UIViewController {

...

 @IBAction func tappedDoneButton(_ sender: Any) {
        guard let navigationViewController = self.presentingViewController as? UINavigationController, let juiceMakerViewController = navigationViewController.viewControllers[0] as? FruitStockViewControllerDelegate else {
            return
        }
        juiceMakerViewController.didDismiss(self)
        self.dismiss(animated: true)
    }
    
}
```

### 시도 결과
`didDismiss`와 `presentedViewControllerDidDismissed`가 호출되고 리턴되지만 `presentedViewDismissed: 
       self.rx.methodInvoked(#selector(JuiceMakerViewController.presentedViewControllerDidDismissed)).map{_ in}`의 시퀀스는 생성되지 않는 것을 확인했습니다...왜 안불리는 걸까요...ㅠㅠ

음... 이 로직도 viewControllers의 배열에 직접 접근한다는 점에서 좋은 해결책은 아니라고 생각이 드네요,,,🥲

이러한 상황에서 줄라이라면 어떻게 대응하실지 의견 여쭤봅니다!
